contracts/governance/BaseWeightedMultisig.sol

```diff
@@ -0,0 +1,185 @@
+        0xa233fbcae4dcfad00091a9d8ff9561f12b3db9ec7227470684b4617d40a38746;
+
+    struct WeightedMultisigStorage {
+        uint256 currentSignersEpoch;
+        mapping(uint256 => bytes32) hashForSignersEpoch;
+        mapping(bytes32 => uint256) signersEpochForHash;
+    }
+
+    uint256 internal constant OLD_SIGNERS_RETENTION = 16;
+
+    /**********************\
+    |* External Functions *|
+    \**********************/
+
+    /*
+     * @notice This function returns the current signers epoch
+     * @return uint256 The current signers epoch
+     */
+    function currentSignersEpoch() external view returns (uint256) {
+        return _baseWeightedStorage().currentSignersEpoch;
+    }
+
+    /*
+     * @notice This function returns the signers hash for a given epoch
+     * @param epoch The given epoch
+     * @return bytes32 The signers hash for the given epoch
+     */
+    function hashForSignersEpoch(uint256 epoch) external view returns (bytes32) {
+        return _baseWeightedStorage().hashForSignersEpoch[epoch];
+    }
+
+    /*
+     * @notice This function returns the epoch for a given signers hash
+     * @param hash The signers hash
+     * @return uint256 The epoch for the given signers hash
+     */
+    function signersEpochForHash(bytes32 hash) external view returns (uint256) {
+        return _baseWeightedStorage().signersEpochForHash[hash];
+    }
+
+    /*************************\
+    |* Integration Functions *|
+    \*************************/
+    /*
+     * @notice This function takes messageHash and proof data and reverts if proof is invalid
+     * @param messageHash The hash of the message that was signed
+     * @param weightedSigners The weighted signers data
+     * @param signatures The signatures data
+     * @return isLatestSigners True if provided signers are the current ones
+     */
+    function _validateProof(
+        bytes32 messageHash,
+        bytes calldata weightedSigners,
+        bytes[] calldata signatures
+    ) internal view returns (bool isLatestSigners) {
+        WeightedMultisigStorage storage $ = _baseWeightedStorage();
+        bytes32 signersHash = keccak256(weightedSigners);
+        uint256 signersEpoch = $.signersEpochForHash[signersHash];
+        uint256 currentEpoch = $.currentSignersEpoch;
+
+        isLatestSigners = signersEpoch == currentEpoch;
+
+        if (signatures.length == 0) revert MalformedSignatures();
+        if (signersEpoch == 0 || currentEpoch - signersEpoch >= OLD_SIGNERS_RETENTION || weightedSigners.length == 0) revert InvalidSigners();
+
+        WeightedSigners memory signers = abi.decode(weightedSigners, (WeightedSigners));
+
+        _validateSignatures(messageHash, signers, signatures);
+    }
+
+    /*
+     * @notice This function rotates the current signers with a new set of signers
+     * @param newWeightedSigners The new weighted signers data
+     */
+    function _rotateSigners(bytes memory newWeightedSigners) internal {
```

Comment:
pass struct directly?

Reply:
Yeah that makes the code cleaner but I now I need to encode them again for hashing. I think it's not too much of an overhead.